README.md
OpenMediaVault plugin to download files using Curl, Aria2, and youtube videos using the yt-dlp fork of youtube-dl.

yt-dlp supplied from: github.com/yt-dlp/yt-dlp


debian/changelog
openmediavault-downloader (7.0) stable; urgency=low

  * Update for OMV 7.x

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 29 Oct 2023 14:49:24 -0500

openmediavault-downloader (6.0.3) stable; urgency=low

  * Remove ulotzo

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 30 Apr 2023 10:44:41 -0500

openmediavault-downloader (6.0.2) stable; urgency=low

  * Improve youtube-dl downloading

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 28 Apr 2023 06:42:28 -0500

openmediavault-downloader (6.0.1) stable; urgency=low

  * Update yt-dlp link

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 12 Apr 2022 06:29:08 -0500

openmediavault-downloader (6.0) stable; urgency=low

  * Update for OMV 6.x

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 19 Dec 2021 12:48:19 -0600

openmediavault-downloader (5.1) stable; urgency=low

  * Add option to download subtitles

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 04 Apr 2020 07:34:33 -0500

openmediavault-downloader (5.0) stable; urgency=low

  * Update for OMV 5.x

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 11 Aug 2019 09:15:09 -0500

openmediavault-downloader (3.6.2) stable; urgency=low

  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 03 Jun 2017 08:46:12 -0500

openmediavault-downloader (3.6.1) stable; urgency=low

  * Remove unneeded line

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 15 May 2017 07:40:42 -0500

openmediavault-downloader (3.6) stable; urgency=low

  * Fix edit for multiple downloads

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 15 May 2017 07:23:45 -0500

openmediavault-downloader (3.5) stable; urgency=low

  * Allow multiple urls to download on one job.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 06 May 2017 09:01:56 -0500

openmediavault-downloader (3.4) stable; urgency=low

  * Improve foreign content handling

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 18 Mar 2017 10:08:17 -0500

openmediavault-downloader (3.3) stable; urgency=low

  * Update for OMV 3.0.40

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 25 Sep 2016 17:21:36 -0500

openmediavault-downloader (3.2) stable; urgency=low

  * Add --ignore-errors to playlist download

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 25 Sep 2016 17:15:22 -0500

openmediavault-downloader (3.1) stable; urgency=low

  * Update for auto notifications

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 07 Jun 2016 20:22:27 -0500

openmediavault-downloader (3.0) stable; urgency=low

  * Update for OMV 3 and datamodels
  * Update youtube-dl to 2016.05.21.2
  * Remove symlinks tab

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 29 May 2016 08:30:14 -0500

openmediavault-downloader (2.1) stable; urgency=low

  * Remove unneeded function
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 31 Aug 2015 07:00:14 -0500

openmediavault-downloader (2.0) stable; urgency=low

  * Add support for downloading playlists
  * Update to youtube-dl 2015.06.25

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 30 Aug 2015 08:51:09 -0500

openmediavault-downloader (1.0.5) stable; urgency=low

  * Fix bug in downloading status
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 09 Dec 2014 10:55:23 -0600

openmediavault-downloader (1.0.4) stable; urgency=low

  * Fix download button translation bug
  * Update youtube-dl to 2014.12.01
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 01 Dec 2014 10:56:40 -0600

openmediavault-downloader (1.0.3) stable; urgency=low

  * Add logging
  * Add downloading and filesize columns
  * Update to youtube-dl 2014.10.13
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 13 Oct 2014 10:34:17 -0500

openmediavault-downloader (1.0.2) stable; urgency=low

  * Fix silent download

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 27 Sep 2014 09:22:51 -0500

openmediavault-downloader (1.0.1) stable; urgency=low

  * Add symlink creation tab
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Thu, 25 Sep 2014 14:48:40 -0500

openmediavault-downloader (1.0) stable; urgency=low

  * Simplify grid toolbar 'selectionchange' event processing. (OMV svn r1327)
  * Update locales
  * Update youtube-dl to 2014.07.20.2

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 21 Jul 2014 09:49:03 -0500

openmediavault-downloader (0.6.4) stable; urgency=low

  * Add button to update youtube-dl
  * Change dependencies for OMV 0.5 without backports enabled
  * Update youtube-dl to 2014.05.19

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 25 May 2014 10:52:14 -0500

openmediavault-downloader (0.6.3) stable; urgency=low

  * Update youtube-dl to 2014.05.05
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 12 May 2014 06:52:06 -0500

openmediavault-downloader (0.6.2) stable; urgency=low

  * Update youtube-dl to 2014.03.30.1
  * Update locales
  * Fix linked field bug

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 30 Mar 2014 13:08:27 -0500

openmediavault-downloader (0.6.1) stable; urgency=low

  * Update youtube-dl to 2014.03.07
  * Update locales
  * Update to work on sardaukar and kralizec
  * Code improvements

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 07 Mar 2014 09:08:27 -0500

openmediavault-downloader (0.6.0) stable; urgency=low

  * Add upload file to shared folder tab
  * Update youtube-dl to 2014.02.10
  * Update locales


usr/share/openmediavault/confdb/create.d/conf.service.downloader.sh
#!/bin/sh
#
# @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
# @author    Volker Theile <volker.theile@openmediavault.org>
# @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
# @copyright Copyright (c) 2009-2013 Volker Theile
# @copyright Copyright (c) 2013-2021 OpenMediaVault Plugin Developers
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

set -e

. /usr/share/openmediavault/scripts/helper-functions

SERVICE_XPATH_NAME="downloader"
SERVICE_XPATH="/config/services/${SERVICE_XPATH_NAME}"

# Versions of the downloader plugins and dependecies
ULOZTO_DOWNLOADER_VERSION="3.3.1"
TFLITE_VERSION="2.11.0"

if ! omv_config_exists "${SERVICE_XPATH}"; then
    echo "Initialize configuration"
    omv_config_add_node "/config/services" "${SERVICE_XPATH_NAME}"
    omv_config_add_key "${SERVICE_XPATH}" "username" ""
    omv_config_add_key "${SERVICE_XPATH}" "uploadref" "0"
    omv_config_add_node "${SERVICE_XPATH}" "downloads" ""
fi

# download latest yt-dlp
/usr/sbin/omv-updateytdl

if [ ! -e "/usr/bin/python" ]; then
  echo "Adding python symlink to python3..."
  ln -s /usr/bin/python3 /usr/bin/python
fi

# install the ulozto-downloader
#pip3 install tflite-runtime==$TFLITE_VERSION ulozto-downloader[auto-captcha]==$ULOZTO_DOWNLOADER_VERSION

exit 0


var/www/openmediavault/js/omv/module/admin/service/downloader/Upload.js
/**
 *
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
 * @copyright Copyright (c) 2009-2013 Volker Theile
 * @copyright Copyright (c) 2013-2020 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/form/Panel.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/form/field/SharedFolderComboBox.js")
// require("js/omv/form/field/UserComboBox.js")

Ext.define("OMV.module.admin.service.downloader.Upload", {
    extend: "OMV.workspace.form.Panel",
    uses: [
        "OMV.Rpc",
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc"
    ],

    rpcService: "Downloader",
    rpcGetMethod: "getUpload",
    rpcSetMethod: "setUpload",

    hideOkButton: true,

    getFormItems: function() {
        var me = this;
        return [{
            xtype: "fieldset",
            title: _("Settings"),
            defaults: {
                labelSeparator: ""
            },
            items: [{
                xtype: "sharedfoldercombo",
                name: "uploadref",
                fieldLabel: _("Shared Folder"),
                plugins: [{
                    ptype: "fieldinfo",
                    text: _("Upload file to this shared folder.")
                }]
            },{
                xtype: "usercombo",
                name: "username",
                fieldLabel: _("File Owner"),
                value: "root"
            },{
                xtype: "button",
                name: "upload",
                text: _("Upload"),
                scope: this,
                handler: Ext.Function.bind(me.onUploadButton, me, [ me ]),
                margin: "0 0 5 0"
            }]
        }];
    },

    onUploadButton: function() {
        var me = this;
        me.doSubmit();
        Ext.create("OMV.window.Upload", {
            title: _("Upload file"),
            service: "Downloader",
            method: "doUpload",
            listeners: {
                scope: me,
                success: function(wnd, response) {
                    me.doReload();
                }
            }
        }).show();
    }
});

OMV.WorkspaceManager.registerPanel({
    id: "upload",
    path: "/service/downloader",
    text: _("Upload"),
    position: 20,
    className: "OMV.module.admin.service.downloader.Upload"
});


var/www/openmediavault/js/omv/module/admin/service/downloader/Symlinks.js
/**
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
 * @copyright Copyright (c) 2009-2013 Volker Theile
 * @copyright Copyright (c) 2013-2014 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/util/Format.js")
// require("js/omv/Rpc.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/window/RootFolderBrowser.js")

Ext.define("OMV.module.admin.service.downloader.Symlink", {
    extend   : "OMV.workspace.window.Form",
    requires : [
        "OMV.workspace.window.plugin.ConfigObject"
    ],
    uses: [
        "OMV.data.Model",
        "OMV.data.Store",
        "OMV.window.RootFolderBrowser"
    ],

    rpcService   : "Downloader",
    rpcGetMethod : "getSymlink",
    rpcSetMethod : "setSymlink",
    plugins      : [{
        ptype : "configobject"
    }],

    width        : 600,

    getFormItems : function() {
        var me = this;
        return [{
            xtype      : "checkbox",
            name       : "enable",
            fieldLabel : _("Enable"),
            checked    : true
        },{
            xtype          : "textfield",
            name           : "source",
            fieldLabel     : _("Source"),
            allowBlank     : false,
            triggers       : {
                folder : {
                    cls     : Ext.baseCSSPrefix + "form-folder-trigger",
                    handler : "onTriggerClick"
                }
            },
            onTriggerClick : function() {
                Ext.create("OMV.window.RootFolderBrowser", {
                    listeners : {
                        scope  : this,
                        select : function(wnd, node, path) {
                            // Set the selected path.
                            this.setValue(path);
                        }
                    }
                }).show();
            }
        },{
            xtype          : "textfield",
            name           : "destination",
            fieldLabel     : _("Destination"),
            allowBlank     : false,
            triggers       : {
                folder : {
                    cls     : Ext.baseCSSPrefix + "form-folder-trigger",
                    handler : "onTriggerClick"
                }
            },
            onTriggerClick : function() {
                Ext.create("OMV.window.RootFolderBrowser", {
                    listeners : {
                        scope  : this,
                        select : function(wnd, node, path) {
                            // Set the selected path.
                            this.setValue(path);
                        }
                    }
                }).show();
            }
        }];
    }
});

Ext.define("OMV.module.admin.service.downloader.Symlinks", {
    extend   : "OMV.workspace.grid.Panel",
    requires : [
        "OMV.Rpc",
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc",
        "OMV.util.Format"
    ],
    uses     : [
        "OMV.module.admin.service.downloader.Symlink"
    ],

    hidePagingToolbar : false,
    stateful          : true,
    stateId           : "a982a76d-6804-1632-b31b-8b48c0ea6dde",
    columns           : [{
        xtype     : "booleaniconcolumn",
        text      : _("Enabled"),
        sortable  : true,
        dataIndex : "enable",
        stateId   : "enable",
        align     : "center",
        width     : 80,
        resizable : false,
        trueIcon  : "switch_on.png",
        falseIcon : "switch_off.png"
    },{
        text      : _("Source"),
        sortable  : true,
        dataIndex : "source",
        stateId   : "source"
    },{
        text      : _("Destination"),
        sortable  : true,
        dataIndex : "destination",
        stateId   : "destination"
    }],

    initComponent: function() {
        var me = this;
        Ext.apply(me, {
            store : Ext.create("OMV.data.Store", {
                autoLoad : true,
                model    : OMV.data.Model.createImplicit({
                    idProperty : "uuid",
                    fields     : [
                        { name : "uuid", type: "string" },
                        { name : "enable", type: "boolean" },
                        { name : "source", type: "string" },
                        { name : "destination", type: "string" }
                    ]
                }),
                proxy    : {
                    type    : "rpc",
                    rpcData : {
                        service : "Downloader",
                        method  : "getSymlinks"
                    }
                }
            })
        });
        me.callParent(arguments);
    },

    onAddButton : function() {
        var me = this;
        Ext.create("OMV.module.admin.service.downloader.Symlink", {
            title     : _("Add symlink"),
            uuid      : OMV.UUID_UNDEFINED,
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton : function() {
        var me = this;
        var record = me.getSelected();
        Ext.create("OMV.module.admin.service.downloader.Symlink", {
            title     : _("Edit symlink"),
            uuid      : record.get("uuid"),
            listeners : {
                scope  : me,
                submit : function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion : function(record) {
        var me = this;
        OMV.Rpc.request({
            scope    : me,
            callback : me.onDeletion,
            rpcData  : {
                service : "Downloader",
                method  : "deleteSymlink",
                params  : {
                    uuid : record.get("uuid")
                }
            }
        });
    }
});

OMV.WorkspaceManager.registerPanel({
    id        : "symlinks",
    path      : "/service/downloader",
    text      : _("Symlinks"),
    position  : 30,
    className : "OMV.module.admin.service.downloader.Symlinks"
});


var/www/openmediavault/js/omv/module/admin/service/downloader/Downloader.js
/**
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
 * @copyright Copyright (c) 2009-2013 Volker Theile
 * @copyright Copyright (c) 2013-2020 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/util/Format.js")
// require("js/omv/form/field/SharedFolderComboBox.js")
// require("js/omv/form/plugin/LinkedFields.js")
// require("js/omv/module/admin/service/downloader/util/Format.js")

Ext.define("OMV.module.admin.service.downloader.Download", {
    extend: "OMV.workspace.window.Form",
    requires: [
        "OMV.workspace.window.plugin.ConfigObject",
        "OMV.form.plugin.LinkedFields"
    ],

    plugins: [{
        ptype: "configobject"
    },{
        ptype: "linkedfields",
        correlations: [{
            conditions: [{
                name: "dltype",
                value: "youtube-dl"
            }],
            name: [
                "format","subtitles"
            ],
            properties: [
                "show"
            ]
        }]
    }],

    rpcService: "Downloader",
    rpcGetMethod: "getDownload",
    rpcSetMethod: "setDownload",

    width: 600,

    getFormItems: function() {
        var me = this;
        return [{
            xtype: "combo",
            name: "dltype",
            fieldLabel: _("Download Type"),
            allowBlank: false,
            editable: false,
            store: [
                [ "aria2", _("aria2") ],
                [ "curl", _("curl") ],
                [ "youtube-dl", _("youtube-dl") ]
            ],
            mode: "local",
            triggerAction: "all",
            value: ""
        },{
            xtype: "textfield",
            name: "filename",
            fieldLabel: _("Filename"),
            allowBlank: false
        },{
            xtype: "combo",
            name: "format",
            fieldLabel: _("Format"),
            allowBlank: false,
            editable: false,
            store: [
                [ "aac", _("aac audio only") ],
                [ "m4a", _("m4a audio only") ],
                [ "mp3", _("mp3 audio only") ],
                [ "mp4", _("mp4/m4a") ],
                [ "wav", _("wav audio only") ],
                [ "webm", _("webm") ]
            ],
            mode: "local",
            triggerAction: "all",
            value: "mp4",
            hidden: true
        },{
            xtype: "textfield",
            name: "url",
            fieldLabel: _("URL"),
            allowBlank: false
        },{
            xtype: "sharedfoldercombo",
            name: "sharedfolderref",
            fieldLabel: _("Shared Folder"),
            readOnly: (me.uuid !== OMV.UUID_UNDEFINED),
            plugins: [{
                ptype: "fieldinfo",
                text: _("Downloads to this shared folder")
            }]
        },{
            xtype: "checkbox",
            name: "subtitles",
            fieldLabel: _("Subtitles"),
            checked: false,
            boxLabel: _("Download subtitles"),
            hidden: true
        },{
            xtype: "checkbox",
            name: "delete",
            fieldLabel: _("Delete"),
            checked: false,
            boxLabel: _("Delete from list of downloads after file is downloaded")
        }];
    }
});

Ext.define("OMV.module.admin.service.downloader.DownloadMultiple", {
    extend: "OMV.workspace.window.Form",
    requires: [
        "OMV.workspace.window.plugin.ConfigObject"
    ],

    plugins: [{
        ptype: "configobject"
    }],

    rpcService: "Downloader",
    rpcGetMethod: "getDownload",
    rpcSetMethod: "setDownload",

    width: 600,

    getFormItems: function() {
        var me = this;
        return [{
            xtype: "combo",
            name: "format",
            fieldLabel: _("Format"),
            allowBlank: false,
            editable: false,
            store: [
                [ "aac", _("aac audio only") ],
                [ "m4a", _("m4a audio only") ],
                [ "mp3", _("mp3 audio only") ],
                [ "mp4", _("mp4/m4a") ],
                [ "wav", _("wav audio only") ],
                [ "webm", _("webm") ]
            ],
            mode: "local",
            triggerAction: "all",
            value: "mp4"
        },{
            xtype: "textarea",
            name: "url",
            fieldLabel: _("URL"),
            allowBlank: false
        },{
            xtype: "sharedfoldercombo",
            name: "sharedfolderref",
            fieldLabel: _("Shared Folder"),
            readOnly: (me.uuid !== OMV.UUID_UNDEFINED),
            plugins: [{
                ptype: "fieldinfo",
                text: _("Downloads to this shared folder")
            }]
        },{
            xtype: "checkbox",
            name: "subtitles",
            fieldLabel: _("Subtitles"),
            checked: false,
            boxLabel: _("Download subtitles")
        },{
            xtype: "checkbox",
            name: "delete",
            fieldLabel: _("Delete"),
            checked: false,
            boxLabel: _("Delete from list of downloads after file is downloaded")
        },{
            xtype: "hiddenfield",
            name: "dltype",
            value: "multiple"
        }];
    }
});

Ext.define("OMV.module.admin.service.downloader.Playlist", {
    extend: "OMV.workspace.window.Form",
    requires: [
        "OMV.workspace.window.plugin.ConfigObject"
    ],

    plugins: [{
        ptype: "configobject"
    }],

    rpcService: "Downloader",
    rpcGetMethod: "getDownload",
    rpcSetMethod: "setDownload",

    width: 600,

    getFormItems: function() {
        var me = this;
        return [{
            xtype: "combo",
            name: "format",
            fieldLabel: _("Format"),
            allowBlank: false,
            editable: false,
            store: [
                [ "aac", _("aac audio only") ],
                [ "m4a", _("m4a audio only") ],
                [ "mp3", _("mp3 audio only") ],
                [ "mp4", _("mp4/m4a") ],
                [ "wav", _("wav audio only") ],
                [ "webm", _("webm") ]
            ],
            mode: "local",
            triggerAction: "all",
            value: "mp4"
        },{
            xtype: "textfield",
            name: "url",
            fieldLabel: _("URL"),
            allowBlank: false
        },{
            xtype: "sharedfoldercombo",
            name: "sharedfolderref",
            fieldLabel: _("Shared Folder"),
            readOnly: (me.uuid !== OMV.UUID_UNDEFINED),
            plugins: [{
                ptype: "fieldinfo",
                text: _("Downloads to this shared folder")
            }]
        },{
            xtype: "checkbox",
            name: "subtitles",
            fieldLabel: _("Subtitles"),
            checked: false,
            boxLabel: _("Download subtitles")
        },{
            xtype: "checkbox",
            name: "delete",
            fieldLabel: _("Delete"),
            checked: false,
            boxLabel: _("Delete from list of downloads after file is downloaded")
        },{
            xtype: "hiddenfield",
            name: "dltype",
            value: "playlist"
        }];
    }
});

Ext.define("OMV.module.admin.service.downloader.Downloads", {
    extend: "OMV.workspace.grid.Panel",
    requires: [
        "OMV.Rpc",
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc",
        "OMV.util.Format"
    ],
    uses: [
        "OMV.module.admin.service.downloader.Download",
        "OMV.module.admin.service.downloader.DownloadMultiple",
        "OMV.module.admin.service.downloader.Playlist"
    ],

    hidePagingToolbar: false,
    autoReload: true,
    stateful: true,
    stateId: "a982a76d-6804-1632-a31b-8b48c0ea6dde",
    columns: [{
        xtype: "textcolumn",
        text: _("Download Type"),
        sortable: true,
        dataIndex: "dltype",
        stateId: "dltype"
    },{
        xtype: "textcolumn",
        text: _("Filename"),
        sortable: true,
        dataIndex: "filename",
        stateId: "filename"
    },{
        xtype: "templatecolumn",
        text: _("URL"),
        sortable: true,
        dataIndex: "url",
        stateId: "url",
        tpl: '<tpl for="url" between="&lt;br/&gt;">{.}</tpl>'
    },{
        xtype: "textcolumn",
        text: _("Shared Folder"),
        sortable: true,
        dataIndex: "sharedfoldername",
        stateId: "sharedfoldername"
    },{
        xtype: "textcolumn",
        text: _("Downloading"),
        sortable: true,
        dataIndex: "downloading",
        stateId: "downloading",
        renderer: function (value) {
            var content;
            if ( value )
                content = _("Yes");
            else
                content = _("No");
            return content;
        }
    },{
        xtype: "textcolumn",
        text: _("Filesize"),
        sortable: true,
        dataIndex: "filesize",
        stateId: "filesize",
        renderer: OMV.module.services.downloader.util.Format.fsRenderer
    },{
        xtype: "textcolumn",
        text: _("Delete after Download"),
        sortable: true,
        dataIndex: "delete",
        stateId: "delete",
        renderer: function (value) {
            var content;
            if ( value )
                content = _("Yes");
            else
                content = _("No");
            return content;
        }
    },{
        xtype: "textcolumn",
        text: _("Format"),
        sortable: true,
        dataIndex: "format",
        stateId: "format"
    }],

    initComponent: function() {
        var me = this;
        Ext.apply(me, {
            store: Ext.create("OMV.data.Store", {
                autoLoad: true,
                model: OMV.data.Model.createImplicit({
                    idProperty: "uuid",
                    fields: [
                        { name: "uuid", type: "string" },
                        { name: "dltype", type: "string" },
                        { name: "filename", type: "string" },
                        { name: "url", type: "string" },
                        { name: "sharedfoldername", type: "string" },
                        { name: "downloading", type: "boolean" },
                        { name: "filesize", type: "string" },
                        { name: "format", type: "string" },
                        { name: "delete", type: "boolean" }
                    ]
                }),
                proxy: {
                    type: "rpc",
                    rpcData: {
                        service: "Downloader",
                        method: "getDownloadList"
                    }
                }
            })
        });
        me.callParent(arguments);
    },

    getTopToolbarItems: function() {
        var me = this;
        var items = me.callParent(arguments);

        Ext.Array.insert(items, 1, [{
            id: me.getId() + "-multiple",
            xtype: "button",
            text: _("Add Multiple"),
            icon: "images/add.png",
            iconCls: Ext.baseCSSPrefix + "btn-icon-16x16",
            handler: Ext.Function.bind(me.onMultipleButton, me, [ me ]),
            scope: me
        },{
            id: me.getId() + "-playlist",
            xtype: "button",
            text: _("Add Playlist"),
            icon: "images/add.png",
            iconCls: Ext.baseCSSPrefix + "btn-icon-16x16",
            handler: Ext.Function.bind(me.onPlaylistButton, me, [ me ]),
            scope: me
        }]);

        Ext.Array.insert(items, 4, [{
            id: me.getId() + "-download",
            xtype: "button",
            text: _("Download"),
            icon: "images/download.png",
            iconCls: Ext.baseCSSPrefix + "btn-icon-16x16",
            handler: Ext.Function.bind(me.onDownloadButton, me, [ me ]),
            scope: me,
            disabled: true
        }]);

        Ext.Array.insert(items, 6, [{
            id: me.getId() + "-update",
            xtype: "button",
            text: _("Update youtube-dl"),
            icon: "images/refresh.png",
            iconCls: Ext.baseCSSPrefix + "btn-icon-16x16",
            handler: Ext.Function.bind(me.onUpdateButton, me, [ me ]),
            scope: me
        }]);

        return items;
    },

    onSelectionChange: function(model, records) {
        var me = this;
        var download = true;
        me.callParent(arguments);
        if(records.length == 1) {
            var record = me.getSelected();
            if(record.get("downloading") == false) {
                download = false;
            }
        }
        me.setToolbarButtonDisabled("download", download);
    },

    onAddButton: function() {
        var me = this;
        Ext.create("OMV.module.admin.service.downloader.Download", {
            title: _("Add download"),
            uuid: OMV.UUID_UNDEFINED,
            listeners: {
                scope: me,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onMultipleButton: function() {
        var me = this;
        Ext.create("OMV.module.admin.service.downloader.DownloadMultiple", {
            title: _("Add multiple downloads"),
            uuid: OMV.UUID_UNDEFINED,
            listeners: {
                scope: me,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onPlaylistButton: function() {
        var me = this;
        Ext.create("OMV.module.admin.service.downloader.Playlist", {
            title: _("Add playlist"),
            uuid: OMV.UUID_UNDEFINED,
            listeners: {
                scope: me,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton: function() {
        var me = this;
        var record = me.getSelected();
        var type = record.get("dltype");
        var windowName = "";
        var title = "";

        switch (type) {
            case "playlist":
                windowName = "OMV.module.admin.service.downloader.Playlist";
                title = _("Edit playlist");
                break;
            case "multiple":
                windowName = "OMV.module.admin.service.downloader.DownloadMultiple";
                title = _("Edit multiple downloads");
                break;
            default:
                windowName = "OMV.module.admin.service.downloader.Download";
                title = _("Edit download");
        }

        Ext.create(windowName, {
            title: _(title),
            uuid: record.get("uuid"),
            listeners: {
                scope: me,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion: function(record) {
        var me = this;
        OMV.Rpc.request({
            scope: me,
            callback: me.onDeletion,
            rpcData: {
                service: "Downloader",
                method: "deleteDownload",
                params: {
                    uuid: record.get("uuid")
                }
            }
        });
    },

    onDownloadButton: function() {
        var me = this;
        var record = me.getSelected();
        OMV.Rpc.request({
            scope: me,
            rpcData: {
                service: "Downloader",
                method: "doDownload",
                params: {
                    uuid: record.get("uuid")
                }
            }
        });
        me.doReload();
        me.setToolbarButtonDisabled("download", true);
    },

    onUpdateButton: function() {
        var me = this;
        var record = me.getSelected();
        OMV.Rpc.request({
            scope: me,
            rpcData: {
                service: "Downloader",
                method: "doUpdate"
            }
        });
    }
});

OMV.WorkspaceManager.registerNode({
    id: "downloader",
    path: "/service",
    text: _("Downloader"),
    icon16: "images/downloader.png",
    iconSvg: "images/downloader.svg"
});

OMV.WorkspaceManager.registerPanel({
    id: "downloads",
    path: "/service/downloader",
    text: _("Downloads"),
    position: 10,
    className: "OMV.module.admin.service.downloader.Downloads"
});
